src/replay/consensus.zig

```diff
@@ -0,0 +1,1243 @@
+        0.67,
+    );
+
+    var expected_slots = try std.ArrayListUnmanaged(Lockout).initCapacity(
+        std.testing.allocator,
+        3,
+    );
+    defer expected_slots.deinit(std.testing.allocator);
+    var lockouts = [_]Lockout{
+        Lockout{ .slot = 3, .confirmation_count = 3 },
+        Lockout{ .slot = 4, .confirmation_count = 2 },
+        Lockout{ .slot = 5, .confirmation_count = 1 },
+    };
+    try expected_slots.appendSlice(std.testing.allocator, &lockouts);
+    replay_tower.last_vote = sig.consensus.vote_transaction.VoteTransaction{
+        .tower_sync = sig.runtime.program.vote.state.TowerSync{
+            .lockouts = expected_slots,
+            .root = null,
+            .hash = Hash.ZEROES,
+            .timestamp = null,
+            .block_id = Hash.ZEROES,
+        },
+    };
+
+    replay_tower.last_vote_tx_blockhash = .non_voting;
+
+    var last_vote_refresh_time: LastVoteRefreshTime = .{
+        .last_refresh_time = sig.time.Instant.now(),
+        .last_print_time = sig.time.Instant.now(),
+    };
+
+    const result = sig.replay.consensus.maybeRefreshLastVote(
+        &replay_tower,
+        &fixture.progress,
+        hash3.slot,
+        &last_vote_refresh_time,
+    );
+
+    try testing.expectEqual(false, result);
+}
+
+test "maybeRefreshLastVote - hotspare validator" {
+    var prng = std.Random.DefaultPrng.init(91);
+    const random = prng.random();
+
+    const root = SlotAndHash{
+        .slot = 0,
+        .hash = Hash.initRandom(random),
+    };
+    const hash3 = SlotAndHash{
+        .slot = 3,
+        .hash = Hash.initRandom(random),
+    };
+    const hash2 = SlotAndHash{
+        .slot = 2,
+        .hash = Hash.initRandom(random),
+    };
+    const hash1 = SlotAndHash{
+        .slot = 1,
+        .hash = Hash.initRandom(random),
+    };
+
+    var fixture = try TestFixture.init(testing.allocator, root);
+    defer fixture.deinit(testing.allocator);
+
+    var trees1 = try std.BoundedArray(TreeNode, MAX_TEST_TREE_LEN).init(0);
+    trees1.appendSliceAssumeCapacity(&[3]TreeNode{
+        .{ hash1, root },
+        .{ hash2, hash1 },
+        .{ hash3, hash2 },
+    });
+
+    try fixture.fill_fork(
+        testing.allocator,
+        .{ .root = root, .data = trees1 },
+    );
+
+    // Update fork stat
+    if (fixture.progress.getForkStats(hash3.slot)) |fork_stat| {
+        fork_stat.*.my_latest_landed_vote = hash2.slot;
+    }
+
+    var replay_tower = try createTestReplayTower(
+        1,
+        0.67,
+    );
+
+    var expected_slots = try std.ArrayListUnmanaged(Lockout).initCapacity(
+        std.testing.allocator,
+        3,
+    );
+    defer expected_slots.deinit(std.testing.allocator);
+    var lockouts = [_]Lockout{
+        Lockout{ .slot = 3, .confirmation_count = 3 },
+        Lockout{ .slot = 4, .confirmation_count = 2 },
+        Lockout{ .slot = 5, .confirmation_count = 1 },
+    };
+    try expected_slots.appendSlice(std.testing.allocator, &lockouts);
```

Comment:
In reference to this, and the two other instances of this in a couple subsequent tests: there's no reason for `lockouts` to be a `var`, and it could very well be inlined.
You could even do something like this:
```suggestion
    var expected_slots: std.ArrayListUnmanaged(Lockout) =
        try .fromOwnedSlice(try std.testing.allocator.dupe(Lockout, &.{
            .{ .slot = 3, .confirmation_count = 3 },
            .{ .slot = 4, .confirmation_count = 2 },
            .{ .slot = 5, .confirmation_count = 1 },
        }));
    defer expected_slots.deinit(std.testing.allocator);
```

Reply:
[b90deb9](https://github.com/Syndica/sig/pull/762/commits/b90deb9ee5133addc8e933914f9a9d6fe0ff84b8)